### PR TITLE
Migrate some `rustc_borrowck` diagnostics to `SessionDiagnostic`

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory

r? @davidtwco

### CODE REVIEW

File: src/test/ui/borrowck/issue-71546.stderr [hunk 1]
```diff
@@ -10,7 +10,7 @@ LL | |         .into_iter()
 LL | |         .map(|elem| elem.to_string())
    | |_____________________________________^
    |
-   = note: could not prove for<'r> [closure@$DIR/issue-71546.rs:14:14: 14:37] well-formed
+   = note: could not prove `for<'r> [closure@$DIR/issue-71546.rs:14:14: 14:37] well-formed`
```
Comment (compiler-errors) [inline src/test/ui/borrowck/issue-71546.stderr:13]: The `Display` impl for `Predicate` means that we either put ticks around this or we don't put ticks around "could not prove Ty: Trait" -- I prefer the backticks around all predicates. 

We could perhaps also avoid this by having `Predicate` have a smarter `Display` impl, or maybe having a special `IntoDiagnosticArg` for `Predicate`.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**